core/src/cluster_info_vote_listener.rs

```diff
@@ -103,82 +99,26 @@ impl SlotVoteTracker {
 pub struct VoteTracker {
     // Map from a slot to a set of validators who have voted for that slot
     slot_vote_trackers: RwLock<HashMap<Slot, Arc<RwLock<SlotVoteTracker>>>>,
-    // Don't track votes from people who are not staked, acts as a spam filter
-    epoch_authorized_voters: RwLock<HashMap<Epoch, Arc<EpochAuthorizedVoters>>>,
-    leader_schedule_epoch: RwLock<Epoch>,
-    current_epoch: RwLock<Epoch>,
-    epoch_schedule: EpochSchedule,
 }
 
 impl VoteTracker {
-    pub fn new(root_bank: &Bank) -> Self {
-        let current_epoch = root_bank.epoch();
-        let vote_tracker = Self {
-            leader_schedule_epoch: RwLock::new(current_epoch),
-            current_epoch: RwLock::new(current_epoch),
-            epoch_schedule: *root_bank.epoch_schedule(),
-            ..VoteTracker::default()
-        };
+    pub(crate) fn new(root_bank: &Bank) -> Self {
+        let vote_tracker = VoteTracker::default();
         vote_tracker.progress_with_new_root_bank(root_bank);
-        assert_eq!(
-            *vote_tracker.leader_schedule_epoch.read().unwrap(),
-            root_bank.get_leader_schedule_epoch(root_bank.slot())
-        );
-        assert_eq!(*vote_tracker.current_epoch.read().unwrap(), current_epoch,);
         vote_tracker
     }
 
-    pub fn get_or_insert_slot_tracker(&self, slot: Slot) -> Arc<RwLock<SlotVoteTracker>> {
-        let mut slot_tracker = self.slot_vote_trackers.read().unwrap().get(&slot).cloned();
-
-        if slot_tracker.is_none() {
-            let new_slot_tracker = Arc::new(RwLock::new(SlotVoteTracker {
-                voted: HashMap::new(),
-                optimistic_votes_tracker: HashMap::default(),
-                voted_slot_updates: None,
-                gossip_only_stake: 0,
-            }));
-            self.slot_vote_trackers
-                .write()
-                .unwrap()
-                .insert(slot, new_slot_tracker.clone());
-            slot_tracker = Some(new_slot_tracker);
-        }
-
-        slot_tracker.unwrap()
+    fn get_or_insert_slot_tracker(&self, slot: Slot) -> Arc<RwLock<SlotVoteTracker>> {
+        let mut slot_vote_trackers = self.slot_vote_trackers.write().unwrap();
```

Comment:
I would prefer to keep the original locking structure to avoid any perf differences (read -> write only in case the slot doesn't exist). 

This is because in the original code each slot should only take the top level `slot_vote_trackers` lock once on creation, whereas in this new code, the top level write lock will be taken for every validator, for every slot. This may affect replay's propagation check which is also reading this lock for specific slots

Reply:
This was changed because the original code has a race condition.
Between the time the read-lock is released and the write-lock is obtained, the object can be modified by a different thread. So always need to check the condition again under the write-lock.

I added an additional look-up under a read-lock, but note that this new code still avoids the race condition above.

Reply:
ah yeah this code was written under the assumption the vote listener thread would be the only one modifying this structure, kind of like the `cluster_slots_service` as well. This does look much cleaner though, thanks!